### PR TITLE
feat: added C++ code to calculate nCr modulo p

### PR DESCRIPTION
**Pull request by Sreshth01**
#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: 
C++ code to calculate nCr modulo p where p is a prime number

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1409"></a>

### CODE REVIEW

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Please enable [GitHub Actions](https://docs.github.com/en/free-pro-team@latest/actions/learn-github-actions) in your repository of this fork in this link: https://github.com/Sreshth01/C-Plus-Plus/actions

**Comment by Panquesito7** [comment]
Please fix the [LGTM](https://lgtm.com/projects/g/TheAlgorithms/C-Plus-Plus/rev/pr-21d684f1bbfa9d2f9985037ff56166cb6b7babea) errors.

**Comment by Sreshth01** [comment]
Ok

**Comment by Sreshth01** [comment]
Done with the required changes, please review it.
Thanks

File: math/ncr_mod_p.cpp [hunk 1]
```diff
@@ -0,0 +1,52 @@
+/**
+ * @file
+ * @brief C++ program to find number of distinct ways of selecting r objects fom
+ * n distinct objects. SInce this number can be very large, we will calulate
+ * answer modulo p
+ */
+#include <iostream>
+#include <vector>
```
Comment (Panquesito7) [inline math/ncr_mod_p.cpp:8]:
```suggestion
#include <iostream>  /// for io operations
#include <vector>   /// for std::vector
```

File: math/ncr_mod_p.cpp [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+std::vector<int64_t> factorial;  // Store pre-computed factorials
+
+/** function to find inverse modulo of given number */
```
Comment (Panquesito7) [inline math/ncr_mod_p.cpp:12]: Missing parameters and return statement documentation.

File: math/ncr_mod_p.cpp [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+int64_t inverse(int64_t i, int64_t m) {
+    if (i == 1) {
+        return 1;
+    }
+
+    return (m - ((m / i) * inverse(m % i, m)) % m + m) % m;
+}
+
+int64_t nCr(int64_t n, int64_t r, int64_t p) {
```
Comment (Panquesito7) [inline math/ncr_mod_p.cpp:21]: Missing function documentation.

File: math/ncr_mod_p.cpp [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        return 0;
+    } else if (r == 0 || r == n) {
+        return 1;
+    } else {
+        return (((factorial[n] * inverse(factorial[r], p)) % p) *
+                inverse(factorial[n - r], p)) %
+               p;
+    }
+}
+/** Main function */
```
Comment (Panquesito7) [inline math/ncr_mod_p.cpp:32]:
```suggestion
/**
 * @brief Main function
 * @returns 0 on exit
 */
```

File: math/ncr_mod_p.cpp [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    for (int i = 1; i <= n; i++) {
+        factorial[i] = (factorial[i - 1] * i) % p;
+    }
+
+    int64_t answer = nCr(n, r, p);
+
+    std::cout << answer;
+
+    return 0;
+}
```
Comment (Panquesito7) [inline math/ncr_mod_p.cpp:52]:
```suggestion
}

```

File: math/ncr_mod_p.cpp [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+        return 1;
+    } else {
+        return (((factorial[n] * inverse(factorial[r], p)) % p) *
+                inverse(factorial[n - r], p)) %
+               p;
+    }
+}
+/** Main function */
+int main() {
+    int n = 0, r = 0, p = 0;
```
Comment (Panquesito7) [inline math/ncr_mod_p.cpp:34]:
```suggestion
    uint16_t n = 0, r = 0, p = 0;
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

File: math/ncr_mod_p.cpp [hunk 7]
```diff
@@ -0,0 +1,68 @@
+/**
+ * @file
+ * @brief C++ program to find number of distinct ways of selecting r objects fom
+ * n distinct objects. SInce this number can be very large, we will calulate
+ * answer modulo p
+ */
+#include <iostream>  /// for io operations
+#include <vector>    /// for std::vector
+
+std::vector<int64_t> factorial;  // Store pre-computed factorials
```
Comment (Panquesito7) [inline math/ncr_mod_p.cpp:10]:
```suggestion
std::vector<int64_t> factorial;  ///< Store pre-computed factorials
```

File: math/ncr_mod_p.cpp [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+/** Function to find inverse modulo of given number
+ * the function takes 2 parameters: i and m
+ * i is the number whose inverse modulo m we want
+ * inverse function is a recurrsive function which calculates inverse modulo m
+ * of i from inverse modulo m of (i modulo m) the return statement retruns i
+ * inverse modulo m using (i modulo m) inverse modulo m
+ */
+int64_t inverse(int64_t i, int64_t m) {
```
Comment (Panquesito7) [inline math/ncr_mod_p.cpp:19]: Missing parameters and return statement documentation.

File: math/ncr_mod_p.cpp [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+/**Function to compute the required nCr.
+ * mathematically nCr = (n!)/(r! x (n-r)!)
+ * for r>n, it is defined to be 0
+ * so to calculate above mathmatical expression , we multiply n! with inverse
+ * modulo of r! and inverse modulo of (n-r)! the function returns the required
+ * nCr mod p
+ */
+int64_t nCr(int64_t n, int64_t r, int64_t p) {
```
Comment (Panquesito7) [inline math/ncr_mod_p.cpp:34]: Same here as above.

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

**Comment by Sreshth01** [comment]
@Panquesito7 can you please explain what do you mean by "parameters and return statement documentation". I have tried explaining about them with function documentation.

**Comment by Panquesito7** [comment]
> what do you mean by "parameters and return statement documentation"

Parameters should be documented using `@param`.
Say the function name is `printSolution`, and it has 1 parameter.

The way to document the function above should be like this:

```c++
      /**
       * @brief Utility function to print array
       * @param arr array where numbers are stored
       * @returns void
       */
       void printSolution(const std::array<32, uint32_t>& arr) {}
```

If you have any other doubts, please tell me and I'll help you. 🙂

File: math/ncr_mod_p.cpp [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+    if (r > n) {
+        return 0;
+    } else if (r == 0 || r == n) {
+        return 1;
+    } else {
+        return (((factorial[n] * inverse(factorial[r], p)) % p) *
+                inverse(factorial[n - r], p)) %
+               p;
+    }
+}
```
Comment (Panquesito7) [inline math/ncr_mod_p.cpp:45]:
```suggestion
}
}  // namespace math

```

File: math/ncr_mod_p.cpp [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+    factorial = std::vector<int64_t>(n + 1);
+
+    factorial[0] = 1;
+
+    // Loop to pre-calculate i factorial for 0<=i<=n
+    for (int i = 1; i <= n; i++) {
+        factorial[i] = (factorial[i - 1] * i) % p;
+    }
+
+    int64_t answer = nCr(n, r, p);
```
Comment (Panquesito7) [inline math/ncr_mod_p.cpp:64]:
```suggestion
    int64_t answer = math::nCr(n, r, p);
```

File: math/ncr_mod_p.cpp [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+
+    std::cout << answer;
```
Comment (Panquesito7) [inline math/ncr_mod_p.cpp:66]:
```suggestion
    std::cout << answer << "\n";
```

File: math/ncr_mod_p.cpp [hunk 13]
```diff
@@ -0,0 +1,69 @@
+/**
+ * @file
+ * @brief C++ program to find number of distinct ways of selecting r objects fom
+ * n distinct objects. SInce this number can be very large, we will calulate
+ * answer modulo p
```
Comment (Panquesito7) [inline math/ncr_mod_p.cpp:5]:
```suggestion
 * answer modulo p
 *
 * @author [Sreshth01](https://github.com/Sreshth01)
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**